### PR TITLE
Feature/EPG GRID rendering

### PR DESCRIPTION
**Pull request by aakashRepoCodes**
##  Description
- Rendering EPG Grid 

## Changes
 - Consuming EPG Api
 -  Rendering EPG Grid
 -  Rendering Taxonomies
 -  Handling Taxonomy click 
 

##  Test Demo
ezgif com-video-to-gif

### CODE REVIEW

**Review by manish218 — COMMENTED** [review]
Looks good. See my inline comments. Let me know if you need clarifications.

File: app/src/main/java/com/spark/fastplayer/common/ViewExtensions.kt [hunk 1]
```diff
@@ -43,4 +49,35 @@ fun Window.activatePlayerPortraitMode() {
         decorView.systemUiVisibility = uiOptions
         clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
     }
-}
\ No newline at end of file
+}
+
+fun Color.Companion.random() : Color {
+    val red = Random.nextInt(256)
```
[the remark this answers is not in the document]
Comment (aakashRepoCodes) [reply app/src/main/java/com/spark/fastplayer/common/ViewExtensions.kt]: Setting it 128 reduces luminance in colors, this makes white text readable in lighter colors as well.

File: app/src/main/java/com/spark/fastplayer/presentation/epg/EPGState.kt [hunk 2]
```diff
@@ -1,16 +1,12 @@
...
 
 import org.openapitools.client.models.EpgRow
+import org.openapitools.client.models.Taxonomy
 
 
 sealed class EPGState {
-
     object Fetch: EPGState()
-
     object Loading: EPGState()
```
[the remark this answers is not in the document]
Comment (aakashRepoCodes) [reply app/src/main/java/com/spark/fastplayer/presentation/epg/EPGState.kt]: Wiped out

File: app/src/main/java/com/spark/fastplayer/presentation/epg/EPGActivity.kt [hunk 3]
```diff
@@ -36,11 +45,14 @@ class EPGActivity : ComponentActivity() {
 
     private val epgViewModel: EPGViewModel by viewModels()
 
+    private var epgState = mutableStateOf<EPGState>(EPGState.Fetch)
+
+    private var showLoading = mutableStateOf(false)
```
[the remark this answers is not in the document]
Comment (aakashRepoCodes) [reply app/src/main/java/com/spark/fastplayer/presentation/epg/EPGActivity.kt]: Done

File: app/src/main/java/com/spark/fastplayer/presentation/epg/EPGActivity.kt [hunk 4]
```diff
@@ -49,13 +61,23 @@ class EPGActivity : ComponentActivity() {
 
         setContent {
             FastPlayerTheme {
-                RenderPlayer()
+                showLoading(showLoading.value)
```
[the remark this answers is not in the document]
Comment (aakashRepoCodes) [reply app/src/main/java/com/spark/fastplayer/presentation/epg/EPGActivity.kt]: `showLoading` now consumes EPGState.

File: app/src/main/java/com/spark/fastplayer/presentation/epg/EPGViewModel.kt [hunk 5]
```diff
@@ -47,4 +50,17 @@ class EPGViewModel @Inject constructor(
...
         }
     }
+
+    private fun filterTaxonomies(epgList: List<EpgRow>) {
+        val  set = HashSet<Taxonomy?> ()
+        viewModelScope.launch(coroutineContextProvider.io) {
+            val map: Map<Taxonomy?, List<EpgRow>> = epgList.groupBy { epgRow ->
+               val tx =  epgRow.programs?.firstOrNull()?.taxonomies?.firstOrNull()
+                set.add(tx)
+                return@groupBy tx
```
[the remark this answers is not in the document]
Comment (aakashRepoCodes) [reply app/src/main/java/com/spark/fastplayer/presentation/epg/EPGViewModel.kt]: Done

**Review by manish218 — APPROVED** [review]

**Pull request merged**